File: core/src/nl/t64/game/rpg/screens/inventory/equipslot/EquipSlotsTables.java
package nl.t64.game.rpg.screens.inventory.equipslot;

import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.components.party.PartyContainer;
import nl.t64.game.rpg.screens.inventory.InventoryUtils;
import nl.t64.game.rpg.screens.inventory.WindowSelector;
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlot;
import nl.t64.game.rpg.screens.inventory.tooltip.ItemSlotTooltip;

import java.util.HashMap;
import java.util.Map;


public class EquipSlotsTables implements WindowSelector {

    private final ItemSlotTooltip tooltip;
    private final Map<String, EquipSlotsTable> equipSlots;

    public EquipSlotsTables(ItemSlotTooltip tooltip) {
        this.tooltip = tooltip;
        this.equipSlots = new HashMap<>(PartyContainer.MAXIMUM);
        this.fillEquipSlotsTables();
        this.setCurrentByIndex(7);
    }

    @Override
    public void setKeyboardFocus(Stage stage) {
        stage.setKeyboardFocus(getCurrentEquipTable());
    }

    @Override
    public ItemSlot getCurrentSlot() {
        return getCurrentEquipSlots().getCurrentSlot();
    }

    @Override
    public ItemSlotTooltip getCurrentTooltip() {
        return tooltip;
    }

    @Override
    public void deselectCurrentSlot() {
        getCurrentEquipSlots().deselectCurrentSlot();
    }

    @Override
    public void selectCurrentSlot() {
        getCurrentEquipSlots().selectCurrentSlot();
    }

    @Override
    public void doAction() {
        getCurrentEquipSlots().dequipItem();
    }

    @Override
    public void hideTooltip() {
        tooltip.hide();
    }

    public int getIndexOfCurrentSlot() {
        return getCurrentEquipSlots().getIndexOfCurrentSlot();
    }

    public void setCurrentByIndex(int index) {
        getCurrentEquipSlots().setCurrentByIndex(index);
    }

    public Table getCurrentEquipTable() {
        return getCurrentEquipSlots().container;
    }

    public EquipSlotsTable getCurrentEquipSlots() {
        return equipSlots.get(InventoryUtils.getSelectedHeroId());
    }

    private void fillEquipSlotsTables() {
        Utils.getGameData().getParty().getAllHeroes()
             .forEach(hero -> equipSlots.put(hero.getId(), new EquipSlotsTable(hero, tooltip)));
    }

}

File: core/src/nl/t64/game/rpg/screens/inventory/equipslot/EquipSlotsTables.kt
package nl.t64.game.rpg.screens.inventory.equipslot

import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.Table
import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.components.party.PartyContainer
import nl.t64.game.rpg.screens.inventory.InventoryUtils
import nl.t64.game.rpg.screens.inventory.WindowSelector
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlot
import nl.t64.game.rpg.screens.inventory.tooltip.ItemSlotTooltip


class EquipSlotsTables(private val tooltip: ItemSlotTooltip) : WindowSelector {

    private val equipSlots: MutableMap<String, EquipSlotsTable> = HashMap(PartyContainer.MAXIMUM)

    init {
        fillEquipSlotsTables()
        setCurrentByIndex(7)
    }

    override fun setKeyboardFocus(stage: Stage) {
        val container: Table = getCurrentEquipTable()
        stage.keyboardFocus = container
        InventoryUtils.setWindowSelected(container)
    }

    override fun getCurrentSlot(): ItemSlot {
        return getCurrentEquipSlots().getCurrentSlot()
    }

    override fun getCurrentTooltip(): ItemSlotTooltip {
        return tooltip
    }

    override fun deselectCurrentSlot() {
        getCurrentEquipSlots().deselectCurrentSlot()
        InventoryUtils.setWindowDeselected(getCurrentEquipTable())
    }

    override fun selectCurrentSlot() {
        getCurrentEquipSlots().selectCurrentSlot()
    }

    override fun hideTooltip() {
        tooltip.hide()
    }

    override fun doAction() {
        getCurrentEquipSlots().dequipItem()
    }

    fun getIndexOfCurrentSlot(): Int =
        getCurrentEquipSlots().getIndexOfCurrentSlot()

    fun setCurrentByIndex(index: Int) {
        getCurrentEquipSlots().setCurrentByIndex(index)
    }

    fun getCurrentEquipTable(): Table {
        return getCurrentEquipSlots().container
    }

    fun getCurrentEquipSlots(): EquipSlotsTable =
        equipSlots[InventoryUtils.getSelectedHeroId()]!!

    private fun fillEquipSlotsTables() {
        gameData.party.getAllHeroes().forEach { equipSlots[it.id] = EquipSlotsTable(it, tooltip) }
    }

}
